lib/Workflow/Condition.pm

```diff
@@ -61,14 +62,51 @@ sub evaluate_condition {
         $condition = $wf->_factory()
             ->get_condition( $orig_condition, $wf->type );
         $log->debug( "Evaluating condition '$orig_condition'" );
-        my $return_value = $condition->evaluate($wf);
+
+        my $return_value;
+        my $result = $condition->evaluate($wf);
+        if (ref $result eq 'Workflow::Condition::IsTrue') {
+            $log->info( "Got true result with '$result' on '$orig_condition'");
+            $return_value = 1;
+        } elsif (ref $result eq 'Workflow::Condition::IsFalse') {
+            $log->info( "Got false result with '$result' on '$orig_condition'");
+            $return_value = 0;
+        } else {
+            $log->fatal( "Evaluate on '$orig_condition' did not return a valid result object" );
+            $log->trace( 'Eval result was ' . Dumper $result );
+            croak "Evaluate on '$orig_condition' did not return a valid result object";
+        }
+
         $wf->{'_condition_result_cache'}->{$orig_condition} = $return_value;
 
         return $return_value;
     }
 }
 
+package Workflow::Condition::Result;
```

Comment:
In our code base, there are 7 locations where we return a value without a message; there are 4 locations where we include a message. What would that be like in the OpenXPKI code base? Is the removal of GreedyOr and CheckReturn bigger than the creation of these return types?